org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/internal/JavaDebugServer.java

```diff
@@ -0,0 +1,132 @@
+/*******************************************************************************
+* Copyright (c) 2017 Microsoft Corporation and others.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Eclipse Public License v1.0
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v10.html
+*
+* Contributors:
+*     Microsoft Corporation - initial API and implementation
+*******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.internal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+import org.eclipse.jdt.ls.core.internal.debug.IDebugServer;
+import org.eclipse.jdt.ls.debug.internal.adapter.DebugSession;
+import org.eclipse.jdt.ls.debug.internal.adapter.DispatcherProtocol;
+import org.eclipse.jdt.ls.debug.internal.adapter.DispatcherProtocol.IResponder;
+import org.eclipse.jdt.ls.debug.internal.adapter.Events.DebugEvent;
+import org.eclipse.jdt.ls.debug.internal.adapter.JsonUtils;
+import org.eclipse.jdt.ls.debug.internal.adapter.Results.DebugResult;
+import org.eclipse.jdt.ls.debug.internal.adapter.Results.ErrorResponseBody;
+import org.eclipse.jdt.ls.debug.internal.adapter.Types.Message;
+import org.eclipse.jdt.ls.debug.internal.core.log.Logger;
+
+import com.google.gson.JsonObject;
+
+public class JavaDebugServer implements IDebugServer {
+	private ServerSocket serverSocket = null;
+	private DebugSession debugSession = null;
+
+	public JavaDebugServer() {
+		try {
+			this.serverSocket = new ServerSocket(0);
+		} catch (IOException e) {
+			Logger.logError(e);
+		}
+	}
+
+	@Override
+	public int getPort() {
+		if (this.serverSocket != null) {
+			return this.serverSocket.getLocalPort();
+		}
+		return -1;
+	}
+
+	@Override
+	public void execute() {
+		if (this.serverSocket != null) {
+			// Execute eventLoop in a new thread.
+			new Thread(new Runnable() {
+
+				@Override
+				public void run() {
+					int serverPort = -1;
+					try {
+						// It's blocking here to waiting for incoming socket connection. 
+						Socket clientSocket = serverSocket.accept();

```

Comment:
I believe once the connection is accepted we can close the serverSocket unless we want to accept more that one connection on the same port.

Later when we disconnect we can must the socket.